src/main/java/com/cloudogu/scm/tracemonitor/TraceStore.java

```diff
@@ -24,60 +24,80 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Striped;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 import org.apache.shiro.SecurityUtils;
-import sonia.scm.store.ConfigurationEntryStoreFactory;
 import sonia.scm.store.DataStore;
+import sonia.scm.store.DataStoreFactory;
 import sonia.scm.trace.SpanContext;
 
 import javax.inject.Inject;
+import javax.inject.Singleton;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElement;
 import javax.xml.bind.annotation.XmlRootElement;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
-import java.util.stream.Collectors;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.function.Supplier;
 
+@Singleton
 public class TraceStore {
 
   private static final String STORE_NAME = "trace-monitor";
-  private final ConfigurationEntryStoreFactory storeFactory;
+  private final DataStoreFactory storeFactory;
   private final GlobalConfigStore globalConfigStore;
+  private final Striped<ReadWriteLock> locks = Striped.readWriteLock(10);
 
   @Inject
-  public TraceStore(ConfigurationEntryStoreFactory storeFactory, GlobalConfigStore globalConfigStore) {
+  public TraceStore(DataStoreFactory storeFactory, GlobalConfigStore globalConfigStore) {
     this.storeFactory = storeFactory;
     this.globalConfigStore = globalConfigStore;
   }
 
   public Collection<SpanContext> getAll() {
     SecurityUtils.getSubject().checkPermission("traceMonitor:read");
-    DataStore<StoreEntry> store = createStore();
     List<SpanContext> spans = new ArrayList<>();
-    store.getAll().values().forEach(entry -> spans.addAll(entry.getSpans()));
+    doSynchronized("", false, () -> {
+      createStore()
+        .getAll()
+        .values()
+        .forEach(entry -> spans.addAll(entry.getSpans()));
+      return null;
+    });
     return spans;
   }
 
   public Collection<SpanContext> get(String kind) {
-    return getAll().stream()
-      .filter(span -> kind.equalsIgnoreCase(span.getKind()))
-      .collect(Collectors.toList());
-  }
+    SecurityUtils.getSubject().checkPermission("traceMonitor:read");
 
-  void add(SpanContext spanContext) {
-    DataStore<StoreEntry> store = createStore();
-    StoreEntry storeEntry = store.get(spanContext.getKind());
-    int configuredStoreSize = globalConfigStore.get().getStoreSize();
-    if (storeEntry == null) {
-      storeEntry = new StoreEntry(configuredStoreSize);
-    }
-    storeEntry = resizeStoreEntryMaxSize(storeEntry, configuredStoreSize);
-    storeEntry.getSpans().add(spanContext);
-    store.put(spanContext.getKind(), storeEntry);
+    final ArrayList<SpanContext> spans = new ArrayList<>();
+    doSynchronized(kind, false, () -> {
+      spans.addAll(createStore().get(kind).getSpans());
+      return null;
+    });
```

Comment:
Why do you copy this into a new `ArrayList`? Wouldn't it be sufficient, to wrap this into `Collections.unmodifiableCollection()` and return it directly?